Cython/Compiler/ExprNodes.py

```diff
@@ -13399,7 +13399,15 @@ def coerce_to_boolean(self, env):
                         self.operand2, env, result_is_bool=True)
                     if operand2 is not self.operand2:
                         self.coerced_operand2 = operand2
-                return self
+                    # Note that cascade.type is only set on success
```

Comment:
Should we set it to `error_type` on failure, so that we always have a type?

Reply:
Most of the time failure means "not optimized" rather than "error", so I'd rather not use `error_type` - I think that could catch us out later.

I could use `py_object_type` since I think failure almost inevitably means that. I'll have an investigate and check that the logic works for that (see if I can work out when `py_object_type` is actually set).

Reply:
We were probably understanding different things here. I think you meant that *here*, we only set `cascade.type` if we can make something reasonable/faster out of it, whereas I previously read your comment as if the attribute was set somewhere externally before and we only had it *available* here if something else succeeded there (which would allow us to set it to `error_type` there and thus always have the `.type` attribute available here, unconditionally).